lib/cereal/postgres/postgres.go

```diff
@@ -626,87 +658,82 @@ func (pg *PostgresBackend) dequeueTask(tx *sql.Tx, taskName string) (int64, *bac
 	} else if err != nil {
 		return 0, nil, errors.Wrap(err, "failed to dequeue task")
 	}
+
 	return tid, task, nil
 }
-func (pg *PostgresBackend) DequeueTask(ctx context.Context, taskName string) (*backend.Task, backend.TaskCompleter, error) {
+
+func (pg *PostgresBackend) DequeueTask(ctx context.Context, workerID uuid4.UUID, taskName string) (*backend.Task, backend.TaskCompleter, error) {
 	ctx, cancel := context.WithCancel(ctx)
 
-	tx, err := pg.db.BeginTx(ctx, nil)
+	tid, task, err := pg.dequeueTask(ctx, workerID, taskName)
 	if err != nil {
 		cancel()
-		return nil, nil, errors.Wrap(err, "failed to dequeue task")
+		return nil, nil, err
 	}
-
 	taskc := &PostgresTaskCompleter{
-		ctx:    ctx,
-		_tx:    tx,
-		cancel: cancel,
+		db:       pg.db,
+		workerID: workerID,
+		tid:      tid,
+		ctx:      ctx,
+		cancel:   cancel,
 	}
 
-	tid, task, err := pg.dequeueTask(tx, taskName)
-	if err != nil {
-		if rerr := tx.Rollback(); rerr != nil {
-			logrus.WithError(rerr).Error("Failed to rollback dequeue task")
-		}
-		cancel()
-		return nil, nil, err
-	}
-	taskc.tid = tid
-
-	// pinger
-	go func() {
-		for {
-			select {
-			case <-taskc.ctx.Done():
-				return
-			case <-time.After(10 * time.Second):
-				err := taskc.useTx(func(tx *sql.Tx) error {
-					_, err := tx.ExecContext(taskc.ctx, "")
-					return err
-				})
-
-				if err != nil {
-					logrus.WithError(err).Error("Transaction pinger failed")
-					return
-				}
-			}
-		}
-	}()
+	go taskc.workerPing()
 
 	return task, taskc, nil
 }
 
-func (taskc *PostgresTaskCompleter) useTx(f func(tx *sql.Tx) error) error {
-	taskc._txLock.Lock()
-	defer taskc._txLock.Unlock()
-	return f(taskc._tx)
+// NOTE(ssd) 2019-07-02: This can keep running even if our actual task
+// executor thread is stuck. This mostly just accounts for completely
+// lost servers.
+func (taskc *PostgresTaskCompleter) workerPing() {
+	for {
+		select {
+		case <-taskc.ctx.Done():
+			return
+		case <-time.After(30 * time.Second):
+			logrus.Debugf("checkin for worker %s", taskc.workerID.String())
+			_, err := taskc.db.ExecContext(taskc.ctx, "UPDATE cereal_busy_task_workers SET last_checkin = NOW() WHERE id = $1", taskc.workerID.String())
+			if err != nil {
+				logrus.WithError(err).Error("failed to update worker check-in time")
```

Comment:
The error might have the detail, but if not we'll probably want to add the workerID to the log output